From ced3b48e9e69bb1c377565984962d24d7754b28a Mon Sep 17 00:00:00 2001
From: Han Xin <hanxin.hx@alibaba-inc.com>
Date: Mon, 27 Sep 2021 15:39:43 +0800
Subject: [PATCH 47/47] crypto: fixup byte_counter reset in copy_pack_data()

When pack_mapped > packed_git_limit, packed_git will use many windows,
the given packed_git may be changed after use_pack(). So we should reset
byte_counter after use_pack().
---
 builtin/pack-objects.c     |  3 +--
 t/t0940/test-1004-clone.sh | 41 ++++++++++++++++++++++++++++++++++++++
 2 files changed, 42 insertions(+), 2 deletions(-)

diff --git a/builtin/pack-objects.c b/builtin/pack-objects.c
index 25ee2aeb75..3060c5822c 100644
--- a/builtin/pack-objects.c
+++ b/builtin/pack-objects.c
@@ -415,8 +415,6 @@ static void copy_pack_data(struct hashfile *f,
 	unsigned long avail;
 	unsigned char decrypt_buffer[4096];
 
-	if (p->cryptor)
-		p->cryptor->byte_counter = offset;
 	while (len) {
 		in = use_pack(p, w_curs, offset, &avail);
 		if (avail > len)
@@ -425,6 +423,7 @@ static void copy_pack_data(struct hashfile *f,
 			hashwrite_try_encrypt(f, in, avail);
 		} else {
 			unsigned long remains = avail;
+			p->cryptor->byte_counter = offset;
 			while (remains) {
 				unsigned long decrypt_size =
 					sizeof(decrypt_buffer);
diff --git a/t/t0940/test-1004-clone.sh b/t/t0940/test-1004-clone.sh
index c5e75b2978..b26229a6b0 100644
--- a/t/t0940/test-1004-clone.sh
+++ b/t/t0940/test-1004-clone.sh
@@ -19,3 +19,44 @@ test_expect_success 'check log of main' '
 	EOF
 	test_cmp expect actual
 '
+
+test_expect_success NEED_GNU_DD 'setup packedgit limited target' '
+	rm -rf target.git &&
+	git init --bare target.git &&
+	git -C target.git config agit.crypto.enabled 1 &&
+	git -C target.git config agit.crypto.secret c2VjcmV0LXRva2VuMTIzNA== &&
+	git -C target.git config agit.crypto.nonce random_nonce &&
+	git -C target.git config core.packedgitlimit 4k &&
+	git -C target.git config pack.packsizelimit 4k
+'
+
+test_expect_success NEED_GNU_DD 'add 1mb blob to target' '
+	rm -rf workdir &&
+	git clone target.git workdir &&
+	(
+		cd workdir &&
+		cat >blob-1m <<-\EOF &&
+		blob-1m, which is bigger than 4k.
+		EOF
+		if type openssl
+		then
+			openssl enc -aes-256-ctr \
+				-pass pass:"$($DD if=/dev/urandom bs=128 count=1 2>/dev/null | base64)" \
+				-nosalt < /dev/zero | $DD bs=1024 count=1024 >>blob-1m
+		else
+			$DD if=/dev/random bs=1024 count=2050 >>blob-1m
+		fi &&
+		git add blob-1m &&
+		test_tick &&
+		git commit -m blob-1m &&
+		git push origin main
+	)
+'
+
+test_expect_success NEED_GNU_DD 'make multi pack in target' '
+	git -C target.git gc
+'
+
+test_expect_success NEED_GNU_DD 'clone from target' '
+	git clone --no-local target.git dest
+'
-- 
patchwork
